File: java/CommDispatcher.java
package com.skydive.java;

import com.skydive.java.data.*;
import com.skydive.java.events.CommEvent;
import com.skydive.java.events.MessageEvent;
import com.skydive.java.events.SignalPayloadEvent;

/**
 * Created by nawba on 16.10.2016.
 */
public class CommDispatcher {

    private byte[] preambleBuffer;
    private byte[] dataBuffer;

    private int preambleBufferCounter;
    private int dataBufferCounter;

    private boolean isPreambleActive;
    private CommMessage.MessageType activePreambleType;

    private int targetDataBufferCounter;

    private int failedReceptionCounter;
    private int successfulReceptionCounter;

    private boolean receivingSignalData;
    private SignalData.Command receivedSignalData;
    private byte[] signalDataBuffer;
    private int signalDataPacketsToReceive;
    private int signalDataPacketsReceived;

    private CommDispatcherListener listener;

    public CommDispatcher(CommDispatcherListener listener) {
        this.listener = listener;
        reset();
    }

    public void proceedReceiving(final byte[] data, final int dataSize) {
        for (int i = 0; i < dataSize; i++) {
            proceedByte(data[i]);
        }
    }

    public void reset() {
        preambleBuffer = new byte[CommMessage.PREAMBLE_SIZE - 1];
        dataBuffer = new byte[CommMessage.MAX_MESSAGE_SIZE];

        preambleBufferCounter = 0;
        dataBufferCounter = 0;

        isPreambleActive = false;
        activePreambleType = CommMessage.MessageType.EMPTY;

        targetDataBufferCounter = 0;

        failedReceptionCounter = 0;
        successfulReceptionCounter = 0;

        receivingSignalData = false;
        receivedSignalData = SignalData.Command.DUMMY;
        signalDataBuffer = null;
        signalDataPacketsToReceive = 0;
        signalDataPacketsReceived = 0;
    }

    private void proceedByte(final byte b) {
        // check for new preamble
        final CommMessage.MessageType newPreamble = updatePreamble(b);
        if (newPreamble != CommMessage.MessageType.EMPTY) {
            if (isPreambleActive) {
                System.out.println("SkyDive::CommDispatcher::FAIL: new preamble " +
                        "received when previous reception not ready");
                failedReceptionCounter++;
            }
            activatePreamble(newPreamble);
            return;
        }

        if (isPreambleActive) {
            // proceed processing data
            dataBuffer[dataBufferCounter] = b;
            dataBufferCounter++;

            if (dataBufferCounter >= targetDataBufferCounter) {
                // enough data in data buffer
                // check signal message condition
                if (activePreambleType == CommMessage.MessageType.SIGNAL
                        && dataBufferCounter == CommMessage.SIGNAL_CONSTRAINT_SIZE) {
                    // command from signal message just received, update target
                    updateTargetDataSizeWithCommand();
                    return;
                }

                // check CRC condition
                if (isValidMessageCrc()) {
                    // data received successfully!
                    successfulReceptionCounter++;
                    if (activePreambleType == CommMessage.MessageType.SIGNAL &&
                            SignalData.hasPayload(SignalData.parseCommand(dataBuffer))) {
                        handleSignalDataPayloadReception();
                        if (isSignalDataComplete()) {
                            receivingSignalData = false;
                            // signal data payload received successfully
                            listener.handleCommEvent(
                                    new SignalPayloadEvent(signalPayloadMessageFactory()));
                        }
                    } else {
                        if (receivingSignalData) {
                            System.out.println("SkyDive::CommDispatcher::FAIL: receiving SignalData not ready");
                            failedReceptionCounter++;
                        }
                        receivingSignalData = false;
                        // preamble message payload received successfully
                        listener.handleCommEvent(
                                new MessageEvent(new CommMessage(activePreambleType, dataBuffer)));
                    }
                } else {
                    // something gone wrong, reset processor
                    System.out.println("SkyDive::CommDispatcher::FAIL: wrong CRC");
                    failedReceptionCounter++;
                }
                deactivatePreamble();
            }
        }
    }

    private CommMessage.MessageType updatePreamble(final byte b) {
        // new preamble is arrived when all preamble bytes are the same
        // and equals to known preamble type
        // and the last received byte is equal to 0
        CommMessage.MessageType result = CommMessage.MessageType.EMPTY;
        if (b == 0) {
            boolean allEquals = true;
            for (byte by : preambleBuffer) {
                if (preambleBuffer[0] != by) {
                    allEquals = false;
                    break;
                }
            }
            if (allEquals) {
                result = CommMessage.getPreambleTypeByChar(preambleBuffer[0]);
            }
        }
        preambleBuffer[preambleBufferCounter] = b;
        preambleBufferCounter++;
        if (preambleBufferCounter >= CommMessage.PREAMBLE_SIZE - 1) {
            preambleBufferCounter = 0;
        }
        return result;
    }

    private void activatePreamble(final CommMessage.MessageType preambleType) {
        isPreambleActive = true;
        activePreambleType = preambleType;

        dataBufferCounter = 0;
        targetDataBufferCounter = CommMessage.getPayloadSizeByType(preambleType);
        if (activePreambleType != CommMessage.MessageType.SIGNAL) {
            targetDataBufferCounter += CommMessage.CRC_SIZE;
        }
    }

    private void updateTargetDataSizeWithCommand() {
        final SignalData.Command command = SignalData.parseCommand(dataBuffer);
        if (SignalData.hasPayload(command)) {
            targetDataBufferCounter += CommMessage.SIGNAL_DATA_PAYLOAD_SIZE + CommMessage.CRC_SIZE;
        } else {
            targetDataBufferCounter += CommMessage.CRC_SIZE;
        }
    }

    private boolean isValidMessageCrc() {
        byte[] tmpArray = new byte[targetDataBufferCounter - CommMessage.CRC_SIZE];
        System.arraycopy(dataBuffer, 0, tmpArray, 0, tmpArray.length);
        final short crcValue = CommMessage.computeCrc16(tmpArray);
        byte crc1 = (byte) (crcValue & 0xff);
        byte crc2 = (byte) ((crcValue >> 8) & 0xff);
        return dataBuffer[targetDataBufferCounter - 2] == crc1
                && dataBuffer[targetDataBufferCounter - 1] == crc2;
    }

    private void initSignalDataPayloadReception(final SignalData.Command command, final short allPackets) {
        receivedSignalData = command;
        signalDataBuffer = new byte[allPackets * CommMessage.SIGNAL_DATA_PAYLOAD_SIZE];
        signalDataPacketsToReceive = allPackets;
        receivingSignalData = true;
        signalDataPacketsReceived = 0;
    }

    private void handleSignalDataPayloadReception() {
        final SignalData.Command command = SignalData.parseCommand(dataBuffer);
        final short allPackets = SignalData.parseAllPacketsNumber(dataBuffer);
        final short packetNumber = SignalData.parseActualPacketNumber(dataBuffer);

        if (!receivingSignalData || receivedSignalData != command) {
            // new data or another data was being received
            initSignalDataPayloadReception(command, allPackets);
        }
        // put data to buffer in to reported position
        System.arraycopy(dataBuffer, CommMessage.SIGNAL_CONSTRAINT_SIZE,
                signalDataBuffer, packetNumber * CommMessage.SIGNAL_DATA_PAYLOAD_SIZE,
                CommMessage.SIGNAL_DATA_PAYLOAD_SIZE);
        signalDataPacketsReceived++;
    }

    private boolean isSignalDataComplete() {
        return signalDataPacketsToReceive == signalDataPacketsReceived;
    }

    private SignalPayloadData signalPayloadMessageFactory() {
        if (receivedSignalData == SignalData.Command.CALIBRATION_SETTINGS_DATA) {
            return new CalibrationSettings(signalDataBuffer);
        } else if (receivedSignalData == SignalData.Command.CONTROL_SETTINGS_DATA) {
            return new ControlSettings(signalDataBuffer);
        } else if (receivedSignalData == SignalData.Command.ROUTE_CONTAINER_DATA) {
            return new RouteContainer(signalDataBuffer);
        } else {
            // error, wrong data command
            return null;
        }
    }

    private void deactivatePreamble() {
        isPreambleActive = false;
        activePreambleType = CommMessage.MessageType.EMPTY;
        dataBufferCounter = 0;
        targetDataBufferCounter = 0;
    }

    public int getFailedReceptionCounter() {
        return failedReceptionCounter;
    }

    public int getSuccessfulReceptionCounter() {
        return successfulReceptionCounter;
    }

    public interface CommDispatcherListener {
        void handleCommEvent(CommEvent event);
    }
}

File: cpp/source/communication/CommDispatcher.cpp
#include "CommDispatcher.hpp"

#include "CalibrationSettings.hpp"
#include "ControlSettings.hpp"
#include "RouteContainer.hpp"
#include "WifiConfiguration.hpp"

#include <string.h>

CommDispatcher::CommDispatcher(void)
{
    reset();
}

CommDispatcher::~CommDispatcher(void)
{
    cleanSignalDataBuffer();
}

void CommDispatcher::reset(void)
{
    preambleBufferCounter = 0;
    dataBufferCounter = 0;

    isPreambleActive = false;
    activePreambleType = IMessage::EMPTY;

    targetDataBufferCounter = 0;

    failedReceptionCounter = 0;
    sucessfullReceptionCounter = 0;

    receivingSignalData = false;
    receivedSignalData = SignalData::DUMMY;
    signalDataBuffer = NULL;
    signalDataPacketsToReceive = 0;
    signalDataPacketsReceived = 0;
}

const IMessage::PreambleType& CommDispatcher::getPreambleType(void) const
{
    return activePreambleType;
}

SignalData::Command CommDispatcher::getCommand(void) const
{
    return SignalData::parseCommand(dataBuffer);
}

SignalData::Parameter CommDispatcher::getParameter(void) const
{
    return SignalData::parseParameter(dataBuffer + IMessage::SIGNAL_COMMAND_SIZE);
}

SignalData CommDispatcher::getSignalData(void) const
{
    return SignalData(dataBuffer);
}

DebugData CommDispatcher::getDebugData(void) const
{
    return DebugData(dataBuffer);
}

ControlData CommDispatcher::getControlData(void) const
{
    return ControlData(dataBuffer);
}

SensorsData CommDispatcher::getSensorsData(void) const
{
    return SensorsData(dataBuffer);
}

AutopilotData CommDispatcher::getAutopilotData(void) const
{
    return AutopilotData(dataBuffer);
}

void CommDispatcher::getSignalDataObject(ISignalPayloadMessage& data)
{
    switch (data.getSignalDataCommand())
    {
    case SignalData::CALIBRATION_SETTINGS:
        reinterpret_cast<CalibrationSettings&>(data) = CalibrationSettings(signalDataBuffer);
        break;
    case SignalData::CONTROL_SETTINGS:
        reinterpret_cast<ControlSettings&>(data) = ControlSettings(signalDataBuffer);
        break;
    case SignalData::ROUTE_CONTAINER:
        reinterpret_cast<RouteContainer&>(data) = RouteContainer(signalDataBuffer);
        break;
    case SignalData::WIFI_CONFIGURATION:
        reinterpret_cast<WifiConfiguration&>(data) = WifiConfiguration(signalDataBuffer);
        break;
    default:
        // error should never reach this point
        break;
    }
    cleanSignalDataBuffer();
}

IMessage* CommDispatcher::retriveMessage(const IMessage::PreambleType preamble,
                                         const IMessage::MessageType expectedControlMessageType)
{
    switch (preamble)
    {
    case IMessage::CONTROL:
        switch (expectedControlMessageType)
        {
        case IMessage::DEBUG_DATA:
            return new DebugData(dataBuffer);

        case IMessage::CONTROL_DATA:
            return new ControlData(dataBuffer);

        case IMessage::SENSORS_DATA:
            return new SensorsData(dataBuffer);

        default: // error
            return NULL;
        }
        break;

    case IMessage::SIGNAL:
        return retriveSignalMessage();

    case IMessage::AUTOPILOT:
        return new AutopilotData(dataBuffer);

    default: // error
        return NULL;
    }
}

IMessage* CommDispatcher::retriveSignalMessage(void)
{
    const SignalData::Command command = getCommand();
    if (SignalData::hasPayload(command))
    {
        IMessage* result;
        switch (command)
        {
        case SignalData::CALIBRATION_SETTINGS_DATA:
            result = new CalibrationSettings(signalDataBuffer);
            break;

        case SignalData::CONTROL_SETTINGS_DATA:
            result = new ControlSettings(signalDataBuffer);
            break;

        case SignalData::ROUTE_CONTAINER_DATA:
            result = new RouteContainer(signalDataBuffer);
            break;

        case SignalData::WIFI_CONFIGURATION_DATA:
            result = new WifiConfiguration(signalDataBuffer);
            break;

        default: // error
            result = NULL;
        }
        cleanSignalDataBuffer();
        return result;
    }
    else
    {
        return new SignalData(dataBuffer);
    }
}

unsigned CommDispatcher::getSucessfullReceptions(void) const
{
    return sucessfullReceptionCounter;
}

unsigned CommDispatcher::getFailedReceptions(void) const
{
    return failedReceptionCounter;
}

void CommDispatcher::clearCounters(void)
{
    sucessfullReceptionCounter = 0;
    failedReceptionCounter = 0;
}

IMessage::PreambleType CommDispatcher::putChar(unsigned char data)
{
    // check for new preamble
    const IMessage::PreambleType newPreamble = updatePreamble(data);
    if (newPreamble != IMessage::EMPTY)
    {
        if (isPreambleActive)
        {
            // new preamble received when previous reception not ready, some fail
            //std::cout << "\nFAIL:  new preamble received when previous reception not ready\n\n";
#ifdef TRACER_H_
            Tracer::Trace("New preamble received when previous reception not ready");
#endif // TRACER_H_
            failedReceptionCounter++;
        }
        activatePreamble(newPreamble);
        return IMessage::EMPTY;
    }

    if (isPreambleActive)
    {
        // proceed processing data
        dataBuffer[dataBufferCounter] = data;
        dataBufferCounter++;

        if (dataBufferCounter >= targetDataBufferCounter)
        {
            // enough data in data buffer

            // check signal message condition
            if (activePreambleType == IMessage::SIGNAL
                    && dataBufferCounter == IMessage::SIGNAL_CONSTRAINT_SIZE)
            {
                // command from signal message just received, update target
                updataTargetDataSizeWithCommand();
                return IMessage::EMPTY;
            }

            // check CRC condition
            if (isValidMessageCrc())
            {
                // data received succesfully!
                sucessfullReceptionCounter++;
                IMessage::PreambleType result = activePreambleType;
                if (activePreambleType == IMessage::SIGNAL &&
                        SignalData::hasPayload(SignalData::parseCommand(dataBuffer)))
                {
                    handleSignalDataPayloadReception();
                    if (!isSignalDataComplete())
                    {
                        result = IMessage::EMPTY;
                    }
                    else
                    {
                        receivingSignalData = false;
                    }
                }
                else
                {
                    if (receivingSignalData)
                    {
                        //std::cout << "\nFAIL: receiving SignalData not ready\n\n";
                        failedReceptionCounter++;
#ifdef TRACER_H_
                        Tracer::Trace("Receiving SignalData not ready");
#endif // TRACER_H_
                    }
                    receivingSignalData = false;
                }
                deactivatePreamble();
                return result;
            }
            else
            {
                // something gone wrong, reset processor
                //std::cout << "\nFAIL: wrong CRC\n\n";
                failedReceptionCounter++;
#ifdef TRACER_H_
                Tracer::Trace("Wrong CRC");
#endif // TRACER_H_
                deactivatePreamble();
                return IMessage::EMPTY;
            }
        }
    }
    return IMessage::EMPTY;
}

IMessage::PreambleType CommDispatcher::updatePreamble(unsigned char data)
{
    // new preamble is arrived when all preamble bytes are the same
    // and equals to known preamlbe type
    // and the last received byte is equal to 0
    IMessage::PreambleType result = IMessage::EMPTY;
    if (data == 0)
    {
        bool allEquals = true;
        for (unsigned i = 0; i < IMessage::PREAMBLE_SIZE - 1; i++)
        {
            if (preambleBuffer[0] != preambleBuffer[i])
            {
                allEquals = false;
                break;
            }
        }
        if (allEquals)
        {
            result = IMessage::getPreabmleTypeByChar(preambleBuffer[0]);
        }
    }

    preambleBuffer[preambleBufferCounter] = data;
    preambleBufferCounter++;
    if (preambleBufferCounter >= IMessage::PREAMBLE_SIZE - 1)
    {
        preambleBufferCounter = 0;
    }

    return result;
}

void CommDispatcher::activatePreamble(IMessage::PreambleType preambleType)
{
    for (unsigned i = 0; i < IMessage::PREAMBLE_SIZE - 1; i++)
    {
        preambleBuffer[0] = 0;
    }

    isPreambleActive = true;
    activePreambleType = preambleType;

    dataBufferCounter = 0;
    targetDataBufferCounter = IMessage::getPayloadSizeByType(preambleType);
    if (activePreambleType != IMessage::SIGNAL)
    {
        targetDataBufferCounter += IMessage::CRC_SIZE;
    }
}

void CommDispatcher::updataTargetDataSizeWithCommand(void)
{
    const SignalData::Command command = SignalData::parseCommand(dataBuffer);
    if (SignalData::hasPayload(command))
    {
        targetDataBufferCounter += IMessage::SIGNAL_DATA_PAYLOAD_SIZE + IMessage::CRC_SIZE;
    }
    else
    {
        targetDataBufferCounter += IMessage::CRC_SIZE;
    }
}

void CommDispatcher::initSignalDataPayloadReception(const SignalData::Command& command, const unsigned short allPackets)
{
    cleanSignalDataBuffer();
    receivedSignalData = command;
    signalDataBuffer = new unsigned char[allPackets * IMessage::SIGNAL_DATA_PAYLOAD_SIZE];
    signalDataPacketsToReceive = allPackets;
    receivingSignalData = true;
    signalDataPacketsReceived = 0;
}

void CommDispatcher::handleSignalDataPayloadReception(void)
{
    const SignalData::Command command = SignalData::parseCommand(dataBuffer);
    const unsigned short allPackets = SignalData::parseAllPacketsNumber(dataBuffer);
    const unsigned short packetNumber = SignalData::parseActualPacketNumber(dataBuffer);

    if (!receivingSignalData || receivedSignalData != command)
    {
        // new data or another data was being received
        initSignalDataPayloadReception(command, allPackets);
    }
    // put data to bufer in to reported position
    memcpy(signalDataBuffer + packetNumber * IMessage::SIGNAL_DATA_PAYLOAD_SIZE,
           dataBuffer + IMessage::SIGNAL_CONSTRAINT_SIZE,
           IMessage::SIGNAL_DATA_PAYLOAD_SIZE);
    signalDataPacketsReceived++;
}

bool CommDispatcher::isSignalDataComplete(void)
{
    return signalDataPacketsToReceive == signalDataPacketsReceived;
}

void CommDispatcher::cleanSignalDataBuffer(void)
{
    if (signalDataBuffer != NULL)
    {
        delete[] signalDataBuffer;
        signalDataBuffer = NULL;
    }
}

bool CommDispatcher::isValidMessageCrc(void) const
{
    const unsigned short crc = IMessage::computeCrc16(dataBuffer, dataBufferCounter - 2);
    unsigned char crc1 = (unsigned char)(crc & 0xff);
    unsigned char crc2 = (unsigned char)((crc >> 8) & 0xff);
    return	dataBuffer[targetDataBufferCounter - 2] == crc1
            && dataBuffer[targetDataBufferCounter - 1] == crc2;
}

void CommDispatcher::deactivatePreamble(void)
{
    isPreambleActive = false;
    activePreambleType = IMessage::EMPTY;
    dataBufferCounter = 0;
    targetDataBufferCounter = 0;
}
